### PR TITLE
Feature/aa 1921 apply commission by credit card to another currency

### PR DESCRIPTION
**Pull request by DaniyarAskarov**
Issue: https://github.com/happy-travel/agent-app-project/issues/1921

Edo consul: 

  "CreditCardOptions": {
    "PaymentsCommission": 2,
    "AvailableCurrencies": ["USD"],
    "DefaultAvailableCurrency": "USD"
  },

### CODE REVIEW

**Review by AlexeyShok — APPROVED** [review]

File: Api/Models/Accommodations/Rate.cs [hunk 1]
```diff
@@ -26,6 +26,12 @@ namespace HappyTravel.Edo.Api.Models.Accommodations
             CreditCardPrice = creditCardPrice;
         }
 
+
+        public Rate(Rate rate, MoneyAmount finalPrice, decimal commission, MoneyAmount creditCardPrice)
```
Comment (DFSko) [inline Api/Models/Accommodations/Rate.cs:30]: Comment

File: Api/Services/Accommodations/Availability/RoomContractSetPriceProcessor.cs [hunk 2]
```diff
@@ -99,9 +99,46 @@ MoneyAmount ChangeProportionally(MoneyAmount price)
                     break;
             }
 
-            var totalFinalPrice = MoneyRounder.Ceil(roomContractSet.Rate.FinalPrice.ApplyCommission(commission));
+            var roomContracts = new List<RoomContract>(roomContractSet.Rooms.Count);
+            for (var i = 0; i < roomContractSet.Rooms.Count; i++)
+            {
+
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/RoomContractSetPriceProcessor.cs:105]: Pls do Ctrl+Alt+L before pull request in RSharp or Rider

File: Api/Services/Accommodations/Availability/RoomContractSetPriceProcessor.cs [hunk 3]
```diff
@@ -99,9 +99,46 @@ MoneyAmount ChangeProportionally(MoneyAmount price)
...
+            var roomContractSetRate = new Rate(roomContractSet.Rate,
+                roomContractSet.Rate.FinalPrice.ApplyCommission(commission),
+                commission,
+                roomContractSet.Rate.FinalPrice.ApplyCommission(creditCardOptions.PaymentsCommission));
+
+            return BuildRoomContractSet(roomContractSet, roomContractSetRate, roomContracts);
+        }
+
+
+        public static List<RoomContractSet> AlignPrices(List<RoomContractSet> sourceRoomContractSets)
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/RoomContractSetPriceProcessor.cs:124]: Comments

File: Api/Services/Accommodations/Availability/RoomContractSetPriceProcessor.cs [hunk 4]
```diff
@@ -99,9 +99,46 @@ MoneyAmount ChangeProportionally(MoneyAmount price)
...
+            {
+                var roomContractSetWithMarkup = AlignPrices(roomContractSet);
+                roomContractSets.Add(roomContractSetWithMarkup);
+            }
+
+            return roomContractSets;
+        }
+
+
+        public static RoomContractSet AlignPrices(RoomContractSet roomContractSet)
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/RoomContractSetPriceProcessor.cs:137]: Comments

File: Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationPriceProcessor.cs [hunk 5]
```diff
@@ -31,31 +31,32 @@ public Task<RoomContractSetAvailability> ApplyMarkups(RoomContractSetAvailabilit
             => _priceProcessor.ConvertCurrencies(availabilityDetails, ProcessPrices, GetCurrency, targetCurrency);
 
 
-        public async Task<RoomContractSetAvailability> AlignPrices(RoomContractSetAvailability value, AgentContext agent)
+        public async Task<RoomContractSetAvailability> AlignPrices(RoomContractSetAvailability value)
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationPriceProcessor.cs:34]: Comments

Comment (suhan3z) [inline Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationPriceProcessor.cs:34]: The name doesn't reveal update of availability

File: Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationPriceProcessor.cs [hunk 6]
```diff
@@ -31,31 +31,32 @@ public Task<RoomContractSetAvailability> ApplyMarkups(RoomContractSetAvailabilit
...
         {
-            var roomContractSet = RoomContractSetPriceProcessor.AlignPrices(value.RoomContractSet,
+            var roomContractSet = RoomContractSetPriceProcessor.AlignPrices(value.RoomContractSet);
+            return UpdatedRoomContractSetAvailability(value, roomContractSet);
+        }
+
+
+        public async Task<RoomContractSetAvailability> ApplyCommission(RoomContractSetAvailability value, AgentContext agent)
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationPriceProcessor.cs:41]: The name doesn't reveal update of availability

File: Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationPriceProcessor.cs [hunk 7]
```diff
@@ -31,31 +31,32 @@ public Task<RoomContractSetAvailability> ApplyMarkups(RoomContractSetAvailabilit
...
-                localityHtId: value.LocalityHtId,
-                evaluationToken: value.EvaluationToken,
-                marketId: value.MarketId,
-                countryCode: value.CountryCode,
-                supplierCode: value.SupplierCode);
+            return UpdatedRoomContractSetAvailability(value, roomContractSet);
         }
 
 
         private static async ValueTask<RoomContractSetAvailability> ProcessPrices(RoomContractSetAvailability value,
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationPriceProcessor.cs:49]: The same

File: Api/Services/Accommodations/Availability/Steps/RoomSelection/IRoomSelectionPriceProcessor.cs [hunk 8]
```diff
@@ -13,6 +13,8 @@ public interface IRoomSelectionPriceProcessor
 
         Task<SingleAccommodationAvailability> ApplyMarkups(SingleAccommodationAvailability response, AgentContext agent);
 
-        Task<SingleAccommodationAvailability> AlignPrices(SingleAccommodationAvailability availabilityDetails, AgentContext agent);
+        Task<SingleAccommodationAvailability> AlignPrices(SingleAccommodationAvailability availabilityDetails);
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/Steps/RoomSelection/IRoomSelectionPriceProcessor.cs:16]: Comment interfaces as well

File: Api/Services/Accommodations/Availability/Steps/RoomSelection/RoomSelectionPriceProcessor.cs [hunk 9]
```diff
@@ -65,34 +56,37 @@ public async Task<SingleAccommodationAvailability> ApplyMarkups(SingleAccommodat
...
-                marketId: response.MarketId,
-                countryCode: response.CountryCode,
-                supplierCode: response.SupplierCode, 
-                supplierAccommodationCode: response.SupplierAccommodationCode);
+            return UpdatedSingleAccommodationAvailability(response, convertedRoomContractSets);
         }
 
 
-        public async Task<SingleAccommodationAvailability> AlignPrices(SingleAccommodationAvailability source, AgentContext agent)
+        public async Task<SingleAccommodationAvailability> AlignPrices(SingleAccommodationAvailability source)
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/Steps/RoomSelection/RoomSelectionPriceProcessor.cs:63]: Comment

File: Api/Services/Accommodations/Availability/Steps/RoomSelection/RoomSelectionPriceProcessor.cs [hunk 10]
```diff
@@ -65,34 +56,37 @@ public async Task<SingleAccommodationAvailability> ApplyMarkups(SingleAccommodat
...
         {
-            var roomContractSets = RoomContractSetPriceProcessor.AlignPrices(source.RoomContractSets,
+            var roomContractSets = RoomContractSetPriceProcessor.AlignPrices(source.RoomContractSets);
+            return UpdatedSingleAccommodationAvailability(source, roomContractSets);
+        }
+
+
+        public async Task<SingleAccommodationAvailability> ApplyCommission(SingleAccommodationAvailability source, AgentContext agent)
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/Steps/RoomSelection/RoomSelectionPriceProcessor.cs:70]: Comment

Comment (suhan3z) [reply Api/Services/Accommodations/Availability/Steps/RoomSelection/RoomSelectionPriceProcessor.cs:70]: The name doesn't reveal update of availability

File: Api/Services/Accommodations/Availability/Steps/RoomSelection/RoomSelectionPriceProcessor.cs (hunk 9 above)
Comment (suhan3z) [reply Api/Services/Accommodations/Availability/Steps/RoomSelection/RoomSelectionPriceProcessor.cs:63]: The name doesn't reveal update of availability

File: Api/Services/Accommodations/Availability/Steps/WideAvailabilitySearch/IWideAvailabilityPriceProcessor.cs [hunk 11]
```diff
@@ -14,6 +14,8 @@ public interface IWideAvailabilityPriceProcessor
 
         Task<Result<List<AccommodationAvailabilityResult>, ProblemDetails>> ConvertCurrencies(List<AccommodationAvailabilityResult> results, Currencies targetCurrency);
 
-        Task<List<AccommodationAvailabilityResult>> AlignPrices(List<AccommodationAvailabilityResult> results, AgentContext agent);
+        Task<List<AccommodationAvailabilityResult>> AlignPrices(List<AccommodationAvailabilityResult> results);
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/Steps/WideAvailabilitySearch/IWideAvailabilityPriceProcessor.cs:17]: Comments

File: Api/Services/Accommodations/Availability/Steps/WideAvailabilitySearch/WideAvailabilityPriceProcessor.cs [hunk 12]
```diff
@@ -82,58 +65,62 @@ public async Task<List<AccommodationAvailabilityResult>> ApplyMarkups(List<Accom
...
-                    countryCode: slimAccommodationAvailability.CountryCode));
+                convertedResults.Add(UpdatedAccommodationAvailabilityResult(slimAccommodationAvailability, convertedRoomContractSets));
             }
 
             return convertedResults;
         }
 
 
-        public async Task<List<AccommodationAvailabilityResult>> AlignPrices(List<AccommodationAvailabilityResult> results, AgentContext agent)
+        public async Task<List<AccommodationAvailabilityResult>> AlignPrices(List<AccommodationAvailabilityResult> results)
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/Steps/WideAvailabilitySearch/WideAvailabilityPriceProcessor.cs:75]: The name doesn't reveal update of availability. And why there's so many nearly the same AlignPrices and ApplyComissions?

File: Api/Services/Accommodations/Availability/Steps/WideAvailabilitySearch/WideAvailabilitySearchService.cs [hunk 13]
```diff
@@ -136,16 +141,24 @@ public async Task<IEnumerable<WideAvailabilityResult>> GetResult(Guid searchId,
                 : Array.Empty<WideAvailabilityResult>();
 
 
-            Task<Result<List<AccommodationAvailabilityResult>, ProblemDetails>> ConvertCurrencies(List<AccommodationAvailabilityResult> availabilityDetails)
-                => _priceProcessor.ConvertCurrencies(availabilityDetails, agent.PreferredCurrency);
+            bool ApplyingAdditionalCommissionNeeded()
```
Comment (suhan3z) [inline Api/Services/Accommodations/Availability/Steps/WideAvailabilitySearch/WideAvailabilitySearchService.cs:144]: thats the 3rd mention of the ApplyingAdditionalCommissionNeeded maybe it should go to the utility method?

**Review by suhan3z — CHANGES_REQUESTED** [review]
According to comments

**Review by suhan3z — APPROVED** [review]

**Pull request merged**